cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+//! A late-stage redundant-reload remover, that runs after registers have been
+//! allocated and stack slots have been given specific offsets.
+
+use crate::cursor::{Cursor, EncCursor};
+use crate::dominator_tree::DominatorTree;
+use crate::ir::{
+    Ebb, Function, Inst, InstBuilder, InstructionData, Opcode, StackSlotData, Value, ValueLoc,
+};
+use crate::isa::{RegInfo, RegUnit, TargetIsa};
+
+use std::vec::Vec;
+
+/// The remover's state.
+pub struct RedundantReloadRemover {
+    // The running environment used to detect redundancies.  This is reinitialised at the start of
+    // each EBB.
+    env: AvailEnv,
+}
+
+impl RedundantReloadRemover {
+    /// Create a new remover state.
+    pub fn new() -> Self {
+        Self {
+            env: AvailEnv::new(0),
+        }
+    }
+
+    /// Clear the state of the remover.
+    pub fn clear(&mut self) {
+        self.env.map.clear();
+        self.env.size = 0;
+    }
+
+    // Initialise the state of the remover.
+    pub fn init(&mut self, size: usize) {
+        self.clear();
+        for _ in 0..size {
+            self.env.map.push(None);
+        }
+        self.env.size = size;
+    }
+
+    /// Run the remover.
+    pub fn run(&mut self, isa: &dyn TargetIsa, func: &mut Function, domtree: &DominatorTree) {
+        let mut ctx = Context {
+            cur: EncCursor::new(func, isa),
+            reginfo: isa.register_info(),
+            domtree: domtree,
+            state: &mut RedundantReloadRemover::new(),
+        };
+        ctx.run()
+    }
+}
+
+struct Context<'a> {
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // Cached ISA information.  We save it here to avoid frequent virtual function calls on the
+    // `TargetIsa` trait object.
+    reginfo: RegInfo,
+
+    // References to contextual data structures we need.
+    domtree: &'a DominatorTree,
+
+    // The running state
+    state: &'a mut RedundantReloadRemover,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        self.cleanup_func();
```

Comment:
Instead, can `cleanup_func` be renamed to `run`? (and let's have the whole block be with the rest of Context's implementation)